beetsplug/tidal.py

```diff
@@ -0,0 +1,863 @@
+                    album, limit=self.config["metadata_search_limit"].get(int)
+                )
+            ]
+        else:
+            candidates += [
+                self._album_to_albuminfo(x)
+                for x in self._search_album(
+                    f"{artist} {album}",
+                    limit=self.config["metadata_search_limit"].get(int),
+                )
+            ]
+
+        self._log.debug(f"_get_lyrics cache: {self._get_lyrics.cache_info()}")
+        self._log.debug(
+            f"_tidal_search cache: {self._tidal_search.cache_info()}"
+        )
+
+        return candidates
+
+    def item_candidates(self, item, artist, album):
+        """Returns TIDAL track candidates for a specific item"""
+        self._log.debug(f"Searching TIDAL for {item}!")
+
+        return (
+            self._search_from_metadata(
+                item, limit=self.config["metadata_search_limit"].as_number()
+            )
+            or []
+        )
+
+    def _album_to_albuminfo(self, album):
+        """Converts a TIDAL album to a beets AlbumInfo
+
+        :param album: An album obtained from the TIDAL API
+        :type album: tidalapi.media.Album
+        :return: A beets AlbumInfo type created with the provided data
+        :rtype: beets.autotag.hooks.AlbumInfo
+        """
+        tracks = []
+
+        # Process tracks
+        # Not using sparse albums as we already have the album
+        # so it's not using up any additional API calls.
+        for track in album.tracks(sparse_album=False):
+            tracks.append(self._track_to_trackinfo(track, album))
+
+        albuminfo = AlbumInfo(
+            album=album.name,
+            album_id=album.id,
+            tidal_album_id=album.id,
+            artist=album.artist.name,
+            artist_id=album.artist.id,
+            tidal_artist_id=album.artist.id,
+            va=len(album.artists) == 1
+            and album.artist.name.lower() == "various artists",
+            mediums=album.num_volumes,
+            data_source=self.data_source,
+            data_url=album.share_url,
+            tracks=tracks,
+            barcode=album.universal_product_number,
+            albumtype=album.type,
+        )
+
+        # Add release date if we have one
+        if album.release_date:
+            albuminfo.year = album.release_date.year
+            albuminfo.month = album.release_date.month
+            albuminfo.day = album.release_date.day
+
+        return albuminfo
+
+    def _track_to_trackinfo(self, track, album=None):
+        """Converts a TIDAL track to a beets TrackInfo
+
+        :param track: A track obtained from the TIDAL API
+        :type track: tidalapi.media.Track
+        :param album: Fills in optional track info, defaults to None
+        :type album: tidalapi.media.Album, optional
+        :return: A beets TrackInfo created with the provided data
+        :rtype: beets.autotag.hooks.TrackInfo
+        """
+        trackinfo = TrackInfo(
```

Comment:
The `track` object also has `popularity` information. Let us save that as well while we are at it.

Reply:
I just added support for popularity information for both tracks and albums.

There are no commands to force sync it at the moment, it grabs the information when tagging files.